vsintegration/src/FSharp.ProjectSystem.FSharp/Project.fs

```diff
@@ -1674,114 +1672,167 @@ namespace rec Microsoft.VisualStudio.FSharp.ProjectSystem
+                    UpdateConfig(pHierProj)
+                    VSConstants.S_OK
+
+                // NOTE: we don't see this being called, at least in common scenarios
                 member x.UpdateProjectCfg_Begin(pHierProj, _pCfgProj, _pCfgSln, _dwAction, pfCancel) =
                     UpdateConfig(pHierProj)
                     VSConstants.S_OK
+
+                // NOTE: we don't see this being called, at least in common scenarios
                 member x.UpdateProjectCfg_Done(pHierProj, _pCfgProj, _pCfgSln, _dwAction, _fSuccess, _fCancel) =
                     UpdateConfig(pHierProj)
+                    projNode.UpdateMSBuildState()
+                    projNode.SetProjectFileDirty(projNode.IsProjectFileDirty)
+                    projNode.ComputeSourcesAndFlags()
                     VSConstants.S_OK
 
             interface IVsUpdateSolutionEvents3 with
+
                 member x.OnBeforeActiveSolutionCfgChange(_oldCfg, _newCfg) =
+
                     // this will be called for each project, but wait dialogs cannot 'stack'
                     // i.e. if a wait dialog is already open, subsequent calls to StartWaitDialog
                     // will not override the current open dialog
-                    waitDialog <-
-                        {
-                            WaitCaption = FSharpSR.GetString FSharpSR.ProductName
-                            WaitMessage = FSharpSR.GetString FSharpSR.UpdatingSolutionConfiguration
-                            ProgressText = None
-                            StatusBmpAnim = null
-                            StatusBarText = None
-                            DelayToShowDialogSecs = 1
-                            IsCancelable = false
-                            ShowMarqueeProgress = true
-                        }
-                        |> WaitDialog.start projNode.Site
-                        |> Some
+                    if waitDialog.IsNone  then 
+                        waitDialog <-
+                            {
+                                WaitCaption = FSharpSR.GetString FSharpSR.ProductName
+                                WaitMessage = FSharpSR.GetString FSharpSR.UpdatingSolutionConfiguration
+                                ProgressText = None
+                                StatusBmpAnim = null
+                                StatusBarText = None
+                                DelayToShowDialogSecs = 1
+                                IsCancelable = false
+                                ShowMarqueeProgress = true
+                            }
+                            |> WaitDialog.start projNode.Site
+                            |> Some
 
                     VSConstants.S_OK
 
                 member x.OnAfterActiveSolutionCfgChange(_oldCfg, _newCfg) =
+
+                    // We have updated the MSBuild state of each project.  Now we can call ComputeSourcesAndFlags
+                    projNode.ComputeSourcesAndFlags()
```

Comment:
Why are we doing this here? This means that the progress dialog is going to close way before all of the ComputeSourcesAndFlags are complete.

We should reference count how many times waitDialog was (going to be) opened, and then only dispose of it when it reaches 0.